src/nvim/terminal.c

```diff
@@ -405,6 +405,10 @@ void terminal_enter(void)
   // erase the unfocused cursor
   invalidate_terminal(s->term, s->term->cursor.row, s->term->cursor.row + 1);
   showmode();
+  // redraw status line so mode() reports terminal mode correctly.
+  // Some status lines display the current mode in the and they might skip a
+  // beat and report the wrong mode (namely normal mode instead of terminal mode).
+  win_redr_status(curwin);
```

Comment:
Can you try moving this to `redraw()`, and confirm whether that fixes the issue still? That would be a more ideal place for this.

```diff
diff --git a/src/nvim/terminal.c b/src/nvim/terminal.c
index d006477c803a..8426f83e4858 100644
--- a/src/nvim/terminal.c
+++ b/src/nvim/terminal.c
@@ -1273,6 +1273,8 @@ static void redraw(bool restore_cursor)
     maketitle();
   }
-
+  win_redr_status(curwin);
+
   if (restore_cursor) {
     ui_cursor_goto(save_row, save_col);
   } else if (term) {
```

Reply:
Seems to work just as good as my attempt to fix the issue.

Yes the other commit tries to fix something else. I'm still trying to isolate a minimal test case to demonstrate the issue without my current config. I ran into it a couple of times without being able to reproduce it reliably but I finally got a lead and though this might be a good place to investigate it since we are already looking at statuslines here ^^. Just ignore it for the time being.

Reply:
I've moved the other commit out into a separate PR.

Reply:
Thanks. So, I think just change this to `redraw_statuslines()` (see https://github.com/neovim/neovim/pull/8347#discussion_r185618695) and if it passes CI I will get it merged.